From e9960eda90e0e6ab1fdf8191486202e83de5e5b9 Mon Sep 17 00:00:00 2001
From: nabijaczleweli <nabijaczleweli@gmail.com>
Date: Mon, 4 May 2020 18:47:07 +0200
Subject: [PATCH 5/6] Fix potential truncation in
 pppoe-server#startPPPDUserMode()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Making the buffer 2*SMALLBUF, we make sure all PppoeOptions
(which are char[SMALLBUF]) can fit alongside the rest of the arguments

Found by GCC 8.3.0:
  gcc -I../ppp -fno-strict-aliasing -Wall -Wstrict-prototypes    '-DPPPOE_PATH="/usr/sbin/pppoe"' '-DPPPD_PATH="pppd"' '-DPLUGIN_PATH="/etc/ppp/plugins/rp-pppoe.so"' '-DPPPOE_SERVER_OPTIONS="/etc/ppp/pppoe-server-options"' -Ilibevent '-DVERSION="3.14"' -c -o pppoe-server.o pppoe-server.c
  pppoe-server.c: In function ‘startPPPDUserMode’:
  pppoe-server.c:1904:80: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size between 210 and 226 [-Wformat-truncation=]
       snprintf(buffer, SMALLBUF, "%s -n -I %s -e %u:%02x:%02x:%02x:%02x:%02x:%02x%s -S '%s'",
                                                                                ^~
  pppoe-server.c:1909:7:
         PppoeOptions, session->serviceName);
       ~~~~~~~~~~~~
  pppoe-server.c:1904:5: note: ‘snprintf’ output 37 or more bytes (assuming 308) into a destination of size 256
       snprintf(buffer, SMALLBUF, "%s -n -I %s -e %u:%02x:%02x:%02x:%02x:%02x:%02x%s -S '%s'",
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         pppoe_path, session->ethif->name,
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         (unsigned int) ntohs(session->sess),
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         session->eth[0], session->eth[1], session->eth[2],
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         session->eth[3], session->eth[4], session->eth[5],
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         PppoeOptions, session->serviceName);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
---
 src/pppoe-server.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/pppoe-server.c b/src/pppoe-server.c
index c0490fb..57f6eec 100644
--- a/src/pppoe-server.c
+++ b/src/pppoe-server.c
@@ -1893,7 +1893,7 @@ startPPPDUserMode(ClientSession *session)
     /* Leave some room */
     char *argv[64];
 
-    char buffer[SMALLBUF];
+    char buffer[2 * SMALLBUF];
 
     int c = 0;
 
@@ -1901,7 +1901,7 @@ startPPPDUserMode(ClientSession *session)
     argv[c++] = "pty";
 
     /* Let's hope service-name does not have ' in it... */
-    snprintf(buffer, SMALLBUF, "%s -n -I %s -e %u:%02x:%02x:%02x:%02x:%02x:%02x%s -S '%s'",
+    snprintf(buffer, sizeof(buffer), "%s -n -I %s -e %u:%02x:%02x:%02x:%02x:%02x:%02x%s -S '%s'",
 	     pppoe_path, session->ethif->name,
 	     (unsigned int) ntohs(session->sess),
 	     session->eth[0], session->eth[1], session->eth[2],
@@ -1916,7 +1916,7 @@ startPPPDUserMode(ClientSession *session)
     argv[c++] = "file";
     argv[c++] = pppoptfile;
 
-    snprintf(buffer, SMALLBUF, "%d.%d.%d.%d:%d.%d.%d.%d",
+    snprintf(buffer, sizeof(buffer), "%d.%d.%d.%d:%d.%d.%d.%d",
 	    (int) session->myip[0], (int) session->myip[1],
 	    (int) session->myip[2], (int) session->myip[3],
 	    (int) session->peerip[0], (int) session->peerip[1],
-- 
2.20.1

